```
Foundation/Decimal.swift:1393:47: error: '+' is unavailable: Please use explicit type conversions or Strideable methods for mixed-type arithmetics.
    exponent = a._exponent - b._exponent - 38 + exponent
                                              ^
Swift.Strideable:5:24: note: '+' was obsoleted in Swift 4
    public static func +(lhs: Self.Stride, rhs: Self) -> Self
                       ^
Foundation/Decimal.swift:1393:28: error: '-' is unavailable: Please use explicit type conversions or Strideable methods for mixed-type arithmetics.
    exponent = a._exponent - b._exponent - 38 + exponent
                           ^
Swift.Strideable:9:24: note: '-' was obsoleted in Swift 4
    public static func -(lhs: Self, rhs: Self) -> Self.Stride
                       ^
```

--- a/lib/Sema/CSGen.cpp
+++ b/lib/Sema/CSGen.cpp
@@ -102,6 +102,7 @@ namespace {
     unsigned haveIntLiteral : 1;
     unsigned haveFloatLiteral : 1;
     unsigned haveStringLiteral : 1;
+    unsigned haveCollectionLiteral : 1;
     
     llvm::SmallSet<TypeBase*, 16> collectedTypes;
 
@@ -120,10 +121,12 @@ namespace {
       haveIntLiteral = false;
       haveFloatLiteral = false;
       haveStringLiteral = false;
+      haveCollectionLiteral = false;
     }
 
     bool haveLiteral() { 
-      return haveIntLiteral || haveFloatLiteral || haveStringLiteral; 
+      return haveIntLiteral || haveFloatLiteral || haveStringLiteral ||
+             haveCollectionLiteral;
     }
   };
 
@@ -225,6 +228,11 @@ namespace {
         }
 
         return { false, expr };
+      }
+
+      if (isa<CollectionExpr>(expr)) {
+        LTI.haveCollectionLiteral = true;
+        return { true, expr };
       }
       
       if (auto UDE = dyn_cast<UnresolvedDotExpr>(expr)) {
@@ -406,73 +414,66 @@ namespace {
     if (lti.collectedTypes.size() == 1) {
       // TODO: Compute the BCT.
 
+      // It's only useful to favor the type instead of
+      // binding it directly to arguments/result types,
+      // which means in case it has been miscalculated
+      // solver can still make progress.
       auto favoredTy = (*lti.collectedTypes.begin())->getWithoutSpecifierType();
-
       CS.setFavoredType(expr, favoredTy.getPointer());
-      
+
       // If we have a chain of identical binop expressions with homogeneous
       // argument types, we can directly simplify the associated constraint
       // graph.
       auto simplifyBinOpExprTyVars = [&]() {
-        if (!lti.haveLiteral()) {
-          for (auto binExp1 : lti.binaryExprs) {
-            for (auto binExp2 : lti.binaryExprs) {
-              if (binExp1 == binExp2)
-                continue;
+        for (auto binExp1 : lti.binaryExprs) {
+          for (auto binExp2 : lti.binaryExprs) {
+            if (binExp1 == binExp2)
+              continue;
 
-              auto fnTy1 = CS.getType(binExp1)->getAs<TypeVariableType>();
-              auto fnTy2 = CS.getType(binExp2)->getAs<TypeVariableType>();
+            auto fnTy1 = CS.getType(binExp1)->getAs<TypeVariableType>();
+            auto fnTy2 = CS.getType(binExp2)->getAs<TypeVariableType>();
 
-              if (!(fnTy1 && fnTy2))
-                return;
+            if (!(fnTy1 && fnTy2))
+              return;
 
-              auto ODR1 = dyn_cast<OverloadedDeclRefExpr>(binExp1->getFn());
-              auto ODR2 = dyn_cast<OverloadedDeclRefExpr>(binExp2->getFn());
+            auto ODR1 = dyn_cast<OverloadedDeclRefExpr>(binExp1->getFn());
+            auto ODR2 = dyn_cast<OverloadedDeclRefExpr>(binExp2->getFn());
 
-              if (!(ODR1 && ODR2))
-                return; 
+            if (!(ODR1 && ODR2))
+              return;
 
-              // TODO: We currently limit this optimization to known arithmetic
-              // operators, but we should be able to broaden this out to
-              // logical operators as well.
-              if (!isArithmeticOperatorDecl(ODR1->getDecls()[0]))
-                return;
+            // TODO: We currently limit this optimization to known arithmetic
+            // operators, but we should be able to broaden this out to
+            // logical operators as well.
+            if (!isArithmeticOperatorDecl(ODR1->getDecls()[0]))
+              return;
 
-              if (ODR1->getDecls()[0]->getBaseName() !=
-                  ODR2->getDecls()[0]->getBaseName())
-                return;
+            if (ODR1->getDecls()[0]->getBaseName() !=
+                ODR2->getDecls()[0]->getBaseName())
+              return;
 
-              // All things equal, we can merge the tyvars for the function
-              // types.
-              auto rep1 = CS.getRepresentative(fnTy1);
-              auto rep2 = CS.getRepresentative(fnTy2);
+            // All things equal, we can merge the tyvars for the function
+            // types.
+            auto rep1 = CS.getRepresentative(fnTy1);
+            auto rep2 = CS.getRepresentative(fnTy2);
 
-              if (rep1 != rep2) {
-                CS.mergeEquivalenceClasses(rep1, rep2, 
+            if (rep1 != rep2) {
+              CS.mergeEquivalenceClasses(rep1, rep2,
+                                         /*updateWorkList*/ false);
+            }
+
+            auto odTy1 = CS.getType(ODR1)->getAs<TypeVariableType>();
+            auto odTy2 = CS.getType(ODR2)->getAs<TypeVariableType>();
+
+            if (odTy1 && odTy2) {
+              auto odRep1 = CS.getRepresentative(odTy1);
+              auto odRep2 = CS.getRepresentative(odTy2);
+
+              // Since we'll be choosing the same overload, we can merge
+              // the overload tyvar as well.
+              if (odRep1 != odRep2)
+                CS.mergeEquivalenceClasses(odRep1, odRep2,
                                            /*updateWorkList*/ false);
-
-                // Since we're merging argument constraints, make sure that
-                // the representative tyvar is properly bound to the argument
-                // type.
-                CS.addConstraint(ConstraintKind::Bind,
-                                 rep1,
-                                 favoredTy,
-                                 CS.getConstraintLocator(binExp1));
-              }
-
-              auto odTy1 = CS.getType(ODR1)->getAs<TypeVariableType>();
-              auto odTy2 = CS.getType(ODR2)->getAs<TypeVariableType>();
-
-              if (odTy1 && odTy2) {
-                auto odRep1 = CS.getRepresentative(odTy1);
-                auto odRep2 = CS.getRepresentative(odTy2);
-
-                // Since we'll be choosing the same overload, we can merge
-                // the overload tyvar as well.
-                if (odRep1 != odRep2)
-                  CS.mergeEquivalenceClasses(odRep1, odRep2,
-                                             /*updateWorkList*/ false);
-              }
             }
           }
         }

--- a/test/Constraints/generics.swift
+++ b/test/Constraints/generics.swift
@@ -496,3 +496,11 @@ extension Sequence {
   func foo24329052<Element>(_ v: Element) { rdar24329052(v) }
   // expected-error@-1 {{cannot convert value of type 'Element' (generic parameter of instance method 'foo24329052(_:)') to expected argument type 'Self.Element' (associated type of protocol 'Sequence')}}
 }
+
+func rdar27700622<E: Comparable>(_ input: [E]) -> [E] {
+  let pivot = input.first!
+  let lhs = input.dropFirst().filter { $0 <= pivot }
+  let rhs = input.dropFirst().filter { $0 > pivot }
+
+  return rdar27700622(lhs) + [pivot] + rdar27700622(rhs) // Ok
+}

--- a/test/Constraints/operator.swift
+++ b/test/Constraints/operator.swift
@@ -152,3 +152,27 @@ let bar_33759839 = ["A", "B", "C"]
 let _: [E_33759839] = foo_33759839.map { .bar($0) } +
                       bar_33759839.map { .bar($0) } +
                       [E_33759839.foo] // Ok
+
+// rdar://problem/28688585
+
+class B_28688585 {
+  var value: Int
+
+  init(value: Int) {
+    self.value = value
+  }
+
+  func add(_ other: B_28688585) -> B_28688585 {
+    return B_28688585(value: value + other.value)
+  }
+}
+
+class D_28688585 : B_28688585 {
+}
+
+func + (lhs: B_28688585, rhs: B_28688585) -> B_28688585 {
+  return lhs.add(rhs)
+}
+
+let var_28688585 = D_28688585(value: 1)
+_ = var_28688585 + var_28688585 + var_28688585 // Ok

--- a/test/Constraints/operator.swift
+++ b/test/Constraints/operator.swift
@@ -138,3 +138,17 @@ func shrinkTooFar(_ : Double, closure : ()->()) {}
 func testShrinkTooFar() {
   shrinkTooFar(0*0*0) {}
 }
+
+// rdar://problem/33759839
+
+enum E_33759839 {
+    case foo
+    case bar(String)
+}
+
+let foo_33759839 = ["a", "b", "c"]
+let bar_33759839 = ["A", "B", "C"]
+
+let _: [E_33759839] = foo_33759839.map { .bar($0) } +
+                      bar_33759839.map { .bar($0) } +
+                      [E_33759839.foo] // Ok

